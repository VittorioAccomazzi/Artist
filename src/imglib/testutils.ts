import { Canvas, createCanvas, loadImage } from 'canvas';
import {imageHash} from 'image-hash'
import * as fs from 'fs'
import * as path from 'path'


export const tmpFolder = 'tmp'
export const testImage1 = 'src/imglib/testimages/test1.jpg'
export const testImage2 = 'src/imglib/testimages/test2.jpg'
export const testImages = [ testImage1, testImage2]

export async function hash( canvas : Canvas) : Promise<string> {
    let b = canvas.toBuffer('image/png');
    let name = `testImage-${canvas.width}x${canvas.height}.png`
    return new Promise((res, rej)=>{
        imageHash({
            ext: 'image/png',
            data: b,
            name: name
        }, 128, true, (err: Error, val : string )=>{
            if( err ) rej(err)
            res(val)
        })
    })
}

export async function* getCanvases() : AsyncGenerator<[Canvas, CanvasRenderingContext2D]>{
    for( let img of testImages) {
        let image = await loadImage(img)
        let canvas= createCanvas(image.width, image.height);
        let ctx = canvas.getContext('2d')
        ctx.drawImage(image, 0, 0)
        yield [canvas, ctx]
    }
}

export async function dumpImage(canvas : Canvas, name? : string ) : Promise<void> {
    let fullname = path.resolve(tmpFolder, name ? name: (await hash(canvas)).substr(0,25))
    if( ! fs.existsSync(tmpFolder)){
        fs.mkdirSync(tmpFolder)
    }
    await saveCanvas(canvas, fullname)
}

export async function saveCanvas( canvas : Canvas, filename : string ) : Promise<void>{
    return new Promise( (res,rej)=>{
        const out = fs.createWriteStream(`${filename}.png`)
        const stream = canvas.createPNGStream()
        stream.pipe(out)
        out.on('finish', () =>  res())
    })
}

export default {}