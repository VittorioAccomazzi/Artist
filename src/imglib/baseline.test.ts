import {testImages,hash, getCanvases} from './testutils'
import {loadImage} from 'canvas'

test('test image shall be present', async ()=>{
    for( let img of testImages) {
        let c = await loadImage(img)
        expect(c).not.toBeNull()
    }
})

test('validate basline hash', async()=>{
    for await ( const [canvas,] of getCanvases() ) {
        let h = await hash(canvas)
        expect(h).toMatchSnapshot();
    }
})

test('validate hash', async()=>{
    for await ( const [canvas,ctx] of getCanvases() ) {
        let h1 = await hash(canvas)
        ctx.beginPath()
        ctx.strokeStyle="FF0A00"
        ctx.fillRect(canvas.width/2, canvas.height/2,20,20)
        ctx.stroke()
        let h2 = await hash(canvas)
        expect(h1).not.toEqual(h2)
    }
}) 
