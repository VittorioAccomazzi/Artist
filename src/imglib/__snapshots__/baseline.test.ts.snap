// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate basline hash 1`] = `"0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000147d9000000000000000000000000005fffffff000000000000000000000004fffffffdf80000000000000000000003fffffffffd200000000000000000007fffffffffffc0000000000000000007fffffffffffff900000000000000003ffffffffffffffff0000000000000007fffffffffffffffff80000000000005ffffffffffffffffffc004000000003ffffffffffffffffffffc0800000000fffffffffffffffffffffff800000000fffffffffffffffffffffff40000007fffffffffffffffffffffffffc00000fffffffffffffffffffffffffff6600bfffffffffffffffffffffffffffd700ffffffffffffffffffffffffffffff90ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffe0fffffffffffffffffffffffffffffff2ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000060000000000000000000000000000003e0000000000000000000003b7b000001c000000000000000000003fffffa0001000000000000000000001ffffffe1f00000000000000000000007fffffffcc1800000000000000000003ffffffffff7e0000000000000000000ffffffffffff00000000000000000007fffffffffff80000000000000000001ffffffffffff80000000000000000001ffffffffffffe0000000000000000007fffffffffffff000000000000000001ffffffffffffffc0000000000000000fffffffffffffffe0000000000000001ffffffffffffffffc000000000000003fffffffffffffffff80000000000001fffffffffffffffffff8000000000003ffffffffffffffffffff000000000007ffffffffffffffffffff00000000001fffffffffffffffffffffe0000000003ffffffffffffffffffffff800000000ffffffffffffffffffffffff00000003ffffffffffffffffffffffffc0000007ffffffffffffffffffffffffc000001ffffffffffffffffffffffffffc0000fffffffffffffffffffffffffef80007ffffffffffffffffffff7fffd000000fffffffffffffffffffe620130000000fffffffffffffffffff2000000000000ffffffffffffffffff80000000000000fffffffffffffc000000000000000000fffffffffff8d0000000000000000000fffffffffffffbffffa0000000000000fffffffffffc0071f070000006000000fffffffffffc005e7048000000000000ffffffffffc000000000000000000000fffff000000000000000000000000000000000000000000000000000000000000000000000000003fffffff80004000000000000000003bfffffffffffffffff0000000012efffffffffffffffffffff000000003fffffffffffffffffffffff00000007ffffffffffffffffffffff7f00000007fffffffffffffffffc53ffff0000002ffffffffffffffffffffffffb101e1ffffffffffffffffffffffe3ffe7817fffffffffffffffffffffc00800183fffffffffffffffffffdf800000000fffffffffffffffffffffffd00000000fffffffffffffffffffffa0000000000ffffffffffffffffffffc20000000000fffffffffffffffffffffff380000000ffffffffffffffffffffe00000000000ffffffffe7ff97ffffffff4000000000ffffffffffdafffe03ffa000000000007fcffffffc3ff81fffdc000000000000dfffff00fff3ffff8000000000000000e7ebff7ffc7ffd7000000000000000001003ffc1ffffb6063f820000000000000000f3efff80007df000000000000000c08406780004f24000000000000000000003ff800ffffb80003b00000000000000000002fff00000020000000000000003801c07f80000018000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffbffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff9fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff4641cfffffffbfffffffffffffffff0ffffffffffff7cbffffffffffff801fffffffe07fffffffffffffffffffffffff803f00087ffffffffff3fffffffeffffffffffe7f87f3e7e1dffbffffffffdf800003ffffffeff90fffffffff7ff800004000ffff001fffffff07381e07d61fffff0c000065e9fe7ffffffdfff38007ff11f3c000003ff02001f8f3f0040c0000007dfc00010001c7fffffffe2000100100000010000000080000fff8bffffdf803ff0000000040000000ffffc000000000000000000187b8000f0000000000007800000008407ffffff000000000000180000000400011fff8034000000061fc0000000000f500000000a300004f9f82000000000000000000001ef80003e00000000000000000000000f70000000000000000000000000000008000000000000000000000000000000000078300000000000000000000000000fff8000300000000000000000000000f000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`validate basline hash 2`] = `"fffe0000a5ffffffbd007fffffc00003fffe0000adfffcdf90003fffffc00000fffe00000bfffcdfc0003fffffe00000fffe00001dffffdfe1001fffffe00010fffe00001ffffe5fe0c01ffffff000e0fffe00000bfdfbcfa0001ffffff001c0fffe0000051bc007c0001ffffff80180fffe000003f9c003c0001ffffffc0100fffe000001bdc003c0001ffffffc0200fffe0000000c8001c0001ffffffe0000fffe00000004800184000ffffffe0000fffe000001bc9ff804000fffffff0000fffe00000298ffff0e000fffffff8000fffe00000903ffffc1000fffffff8000f9fe0000003bffffe0000fffffffc000f9fe000001f1fffff8000fffffff8000f1fe000001037ffffc000fffffff8000f1fe00000003fffffe000fffffff0000e1fe00000202ffffff800ff7fffc0000e1fe000004017fffffc00fe3fffc0001c1fe000000007fffffe00fe3fff8000381fe00000000fffffff00fe1fff0000781fe00000000fffffff80fe0fff0000701fe00000001fffffff80fe0ffe0000f01fe00000003fffffffe0fe07fc0001f01fe0000000fffffffff0fe07fc0003f01fe0000400ffbffffff0fe03f80007f01fe0000800ffbffffff8fe01f0000ff01fe0000801fc7ffffffcfe03f0000ff01fe0001801f8fffffffcfe07f0001ff01fe000180231fffffffefe0ff0003ff00fe000180443fffffffe7e3ff8007ff00fe000180007ffffffff7c7ff8007ff00fe00030000ffffffffff9fff800fff00fe00030001ffffffffffffff000fff00fe00030007ffffffffffffff001fff00fe00038007bfffffffffffff001fff00fe0003800f7fffffffffffff003fff00fe0003801efffffffffffffe007fff00fe0003c029fffffffffffffe007fff00fe0003c053fffffffffffffe00ffff00fe0003c087bffffffffffffc00ffff00fe0003f00f3ff1ffffffffcc01ffff00fe0003f0085c0cfe7fffff0c01ffff00fe0003d8103a0e007ffffe1803ffff00fe0003c0003400007ffff03003ffff00fe0003c860400001ffffc0e007ffff00fe0003d800800002ffff03c007ffff00fe0001e800400005fffc7f800fffff00fe0001f00000000ffff03f000fffff00fe0001f00000001ffff03e000fffff00fe0000f00000000ffff03c0007ffff00fe0000f00000001ffff8300007ffff00fe0000700000003ffff8300007ffff00fe0000600000007ffff830000fffff00fe000060000001fffff830001fffff00fe000060000003fffffc10005fffff00fe000060000007fffffc10005fffff00fe00002000000ffffffc1000bfffff00fe00000400000ffffffc1000bfffff00fe00001000001ffffffc10007fffff00fe00018800003fc1fffc10007ffffe00fe00010000007fb07ff008007ffffc00fe0003800000fe083fc00800fffffc00fe000fc00000fc003fc00803ffffff00fe001f800003f800ff000803ffffff00fe001e000003f00c3f080803ffffff00fe001e000003f00e3e080807ffffff00fe003c004007c67e3e780807ffffff00fe00700000079f977fc00807ffffff00ff01c0c0401f1fffffe0080fffffff00fe010001000f3ffffff0080fffffff00fe03c000401e1fffeff8080fffffff00fe0f0080001c3ffffff80c1fffffff00fe00000000781ffffff80c1ffff7ff00ff01000008383fffdff80c1fffffff00ff06410020703fffcff80c1fffffff00fe18010030601fffcff80c3fffffff00fe04010000c01ffe4ff8043fffffff00fe00010003c01ffe77f0063fffffff00ff00000003801ffe87f0067fffffff00ff00008003801fff03f0067fffffff01ff00008409001fffffe0067fffffff01ff04004032000fffffe0067fffffff01ff08002070001fffffe002ffffffff01ff10000060001fffffc002ffff7fff01ff000011c8001fffffc002ffff7fff01ff080007d00007c0008003ffff7fff01ff480407900007e1ff8003ffff7fff01ff90040f600003f1ff0003ffff7fff01ff00000fc00001f8fe0003ffff7fff00ff200827800001ff1e0003fffe7fff00ff000d03000000fffc0003fffe7fff00ff00011b000001fffc0003fffe7fff00ff0000d8000000fffc0003fffe7fff01ff0008780000003ff80003fffcfffe01ff200c7f0000001ff80003fffffff801ffc0062d800003fffe00031fffffe000ff800664c00007ffff8003c07fffc000ff2002c5800007ffffe003f800ff80c0ff400357a00003fffff803fe00ff00e0ffc1008f100003fffffc03ff84fe00f0ff800107100007fffffe03ffedfc0078ffc00007a80007ffffff07fffffc007cff800003d80003ffffff07fffffc003cff800001c80047ffffff87fffff8003cff800001880007ffffffc7fffff8003cff800001c80007ffffffc7fffff0003cff804009d80047ffffffc7fffff0001cff00000bf000c7ffffffe7fffff0001cff8000073800c7ffffffe7fffff0001cff8000038001c7ffffffe7ffffe0001cff800003c001cffffffff7ffffe0001cff80000720018fffffffffffffc0001cff80000130038ffffffffcffffc0001c7f80000000039ffffffff81fffc0001e7f80000800071ffffffff801ff80001e7f00000c00063ffffffff8007f00001e7f00000600063ffffffff8007f00001e7f800003400c7ffffffffc007e00001e7f8000008010fffffffffc007c00003e7f8000000000fffffffffc007800001e7f0000010003fffffffffe007000000e7f0002010007fffffffffe006000400f7f000101001ffffffffffe020000800f7e000100001ffffffffffe060001800f7f000080003fffffffffff071803000f7e000000007fffffffffff07f80300"`;
